Add support to shield x_nucleo_iks02a1

avisconti opened the pull request:
Add support to shield x_nucleo_iks02a1

Add the board/shield support plus the sample to test it. The samples are created over three different applications: standard (Mode 1), sensorhub (Mode 2), microphone.

The PR has been tested using x_nucleo_iks02a1 shield on top of a Nucleo F401re board (stand and sensorhub) and on top of Nucleo F411re board (microphone).

avisconti committed:
boards/nucleo_f411re: Enable I2S support

I2S may be needed by any shield mounted on this nucleo.
So change pinmux and dts file.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
boards/shields: add support for x-nucleo-iks02a1 shield

x-nucleo-iks02a1 shield is an arduino compatible companion board
which can be used on top of Nucleo standard boards for industrial
applications. It extends the Nucleo functionalities adding following
MEMS sensor support:

    - ISM330DHCX 3-axis accelerometer and 3-axis gyroscope
    - IIS2MDC 3-axis magnetometer
    - IIS2DLPC 3-axis accelerometer
    - IMP34DT05 digital microphone.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
samples/shields: add samples to test x-nucleo-iks02a1 shield

Provide three basic examples to test the x-nucleo-iks02a1 shield:

    - Test shield standard mode
        Acquire sensor data from shield configured in mode 1.

    - Test shield sensorhub mode
        Acquire sensor data from shield configured in mode 2.

    - Test on-shield microphone
        Acquire microphone PDM audio and output it in 16-bit
        PCM format to console.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@erwango 
Can you  give me some feedback specifically to 011e064? 
If you agree we may need to do same to other nucleo board as well.

erwango commented on boards/shields/x_nucleo_iks02a1/x_nucleo_iks02a1_shub.overlay:
```diff
@@ -0,0 +1,22 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&arduino_i2c {
+
+	iis2dlpc@19 {
+		compatible = "st,iis2dlpc";
+		reg = <0x19>;
+		drdy-gpios =  <&arduino_header 4 GPIO_ACTIVE_HIGH>; /* A4 - INT2 */
+		label = "IIS2DLPC";
+	};
+
+	ism330dhcx@6b {
+		compatible = "st,ism330dhcx";
+		reg = <0x6b>;
+		drdy-gpios =  <&arduino_header 11 GPIO_ACTIVE_HIGH>; /* D5 - INT2 */
+		label = "ISM330DHCX";
+	};
+};
```
I know we've split overlay files this way before.
But is that really required? these are the same nodes, at least from device tree point of view.
The only difference is how they'll be configured at the Kconfig level. So, a unique overlay file should be fine IMHO (together with _mic)

avisconti replied:
ok, I think it can be done.
I'll take a look at it right now.

avisconti replied:
OOops, just a couple of doubts that I found immediately:

We have 3 different configurations:
  - standard (Mode 1)
  - sensorhub (Mode 2)
  -  microphone

1. At least the standard and sensorhub solution should have two different dts. This because the IIS2MDC is visible only in standard mode. In sensorhub mode is infact attached to ISM330HCX sensorhub, so it should not be declared in dts. I guess that using a single overlay file would lead to confusion.
2. If I put the i2s part in standard overlay (for example) and use it for microphone sample, I need to be add arduino_i2s for othe board as well, like for example nucleo_f401re, otherwise the sensor standard mode can be tried on f411re only. What do you think?

avisconti replied:
OK, I solved #2 adding arduino_i2s to f401re as well. Other boards may be extended in the future.

For #1 I propose to use two separate overlay files (I removed the mic one). Another possibility is to use ifdefs in the overlay. What do you think?

erwango replied:
> At least the standard and sensorhub solution should have two different dts. This because the IIS2MDC is visible only in standard mode. In sensorhub mode is infact attached to ISM330HCX sensorhub, so it should not be declared in dts. I guess that using a single overlay file would lead to confusion.

I'm not quite in line here. dts is supposed to describe available hardware and this is the same shield you are using in both cases, there is no reason to use 2 different hardware descriptions.
To my understanding hub is a software configuration of the available hardware, this should be done in Kconfig and btw this is already what you are doing with `CONFIG_LSM6DSO_SENSORHUB`.
Are you facing issues when using same overlay in both cases ?


> Another possibility is to use ifdefs in the overlay.

We can't use #ifdef in overlays as dts is processed before Kconfig

avisconti replied:
> > At least the standard and sensorhub solution should have two different dts. This because the IIS2MDC is visible only in standard mode. In sensorhub mode is infact attached to ISM330HCX sensorhub, so it should not be declared in dts. I guess that using a single overlay file would lead to confusion.
> 
> I'm not quite in line here. dts is supposed to describe available hardware and this is the same shield you are using in both cases, there is no reason to use 2 different hardware descriptions.

But the shield can be configured using JP7. There are two modes: Mode1 (standard) and Mode2 (sensorhub). Once that you move that switch in one of the two positions the hardware is different. The connections are different and IIS2MDC is no longer accessible from MCU, but only from ISM330DHCX. So, two different hardwares requires two different DTS.

> To my understanding hub is a software configuration of the available hardware, this should be done in Kconfig and btw this is already what you are doing with `CONFIG_LSM6DSO_SENSORHUB`.

No, no. It is really a different h/w configuration. IIS2MDC is no longer connected to MCU I2C.

> Are you facing issues when using same overlay in both cases ?
> 
> > Another possibility is to use ifdefs in the overlay.
> 
> We can't use #ifdef in overlays as dts is processed before Kconfig
ok. So I don't see other options but having two dts (because we have two different h/w on the same shield)

avisconti replied:
Just one more info here. IIS2MDC sensor is on I2C1 bus, while the other two are on I2C1, where the mcu connects to. 
Using JP7 you may connect I2C2 to I2C1 (Mode 1) so that also IIS2MDC is visible to mcu, or you may connect it to SCx/SDx of ISM330DHCX, so that is ONLY that sensor that can access it.
I find that using the same dts may result in a confusion.

erwango replied:
ok, in that case, that's indeed 2 different hardware combinations and 2 overlays are needed. Thanks for the clarification.
Maybe this is worth a comment directly in the overlay files though.

avisconti replied:
Yes, sure, I'll add it. I will also review the board/shield/iks02a1 index.rst documentation in order to understand if the info are enough.

erwango replied:
Shield doc is clear on that point, but it wouldn't hurt adding a reminder in these overlay files

avisconti replied:
done

erwango commented on samples/shields/x_nucleo_iks02a1/microphone/prj.conf:
```diff
@@ -0,0 +1,17 @@
+# disabled and DMA interrupts must run at maximum priority.
+
+CONFIG_DMA=y
+CONFIG_DMA_0_IRQ_PRI=0
```
Can't this be configured in dt ?

avisconti replied:
This would require a change in the stm32 dma driver and related dts. Actually it is a  bit risky to do it here, as it impacts many applications and would require an exhaustive test.

erwango replied:
ack

avisconti replied:
Well, it was necessary sometime ago, when I did the same for ArgonKey board. Without this definition the audio stream was broken, as the DMA IRQ was not served fast enough. Now I'm relooking again at it and it seems instead that the priority is taken from DTS and it is fixed to 0 by default:

#define DT_ST_STM32_DMA_40026400_IRQ_0_PRIORITY     0

I tried to remove CONFIG_DMA_0_IRQ_PRI from prj.conf and it still works.
I'm not sure where the default priority is set in the dts and how to enforce it.
Any idea? @erwango

avisconti replied:
OK, i think that I found it, sorry.
It is in ./dts/arm/st/f4/stm32f4.dtsi :

interrupts = <56 3 57 3 58 3 59 3 60 3 68 3 69 3 70 3>;

avisconti replied:
Moreover I noticed that CONFIG_DMA and HEAP_SIZE are already set in soc. So I'm gonna delete completely this section.

erwango reviewed (changes requested):
First batch of comments

avisconti commented on boards/shields/x_nucleo_iks02a1/boards/nucleo_f411re.defconfig:
```diff
@@ -0,0 +1,30 @@
+
+if I2S
+
+config I2S_STM32
```
No, it is not. I checked it right now. If I remove I2S_STM32 from here it stay unconfigured and the microphone sample does not work.

I an try to put it in soc (with your help), but I do not see similar stuff done for I2C and SPI for example (I2C_STM32 and SPI_STM32), so I'm a bit confused.

avisconti replied:
Oh, sorry, I misread the soc part.
It seems that I2C_STM32 and SPI_STM32 are enabled in soc/arm/st_stm32/common/Kconfig.defconfig.series.

But I'm not sure I2S should be enabled for all series. Can you help me out in this task?

avisconti replied:
OK, I think it is the right place. The I2S driver Kconfig then depends from I2S (drivers/i2s/Kconfig) and from SOC_SERIES_STM32F4X (drivers/i2s/Kconfig.stm32). So, ultimately it is due to the driver side to filter out the dependencies.

What do you think?
I'm adding it in the soc common part.

erwango replied:
`soc/arm/st_stm32/common/Kconfig.defconfig.series` is the right place indeed.

avisconti commented on boards/shields/x_nucleo_iks02a1/boards/nucleo_f411re.defconfig:
```diff
@@ -0,0 +1,30 @@
+config I2S_STM32_PLLI2S_PLLR
+	default 3
+
+config I2S_1
```
Actually I have some doubts here. I explain to you, I think that which I2S should be enabled does not depend by board (nucleo_f411re does not need any I2S to be enabled because it does not have any audio h/w capability) but by shield. The x_nucleo_f411re shield has a microphone exported onto I2S_1 of the hosting board.

avisconti replied:
If I put it in board I should enable all 5 I2S, which is not good.
Actually It was done like this before, but now it is good that the shield part can overload the defconfig as well. So, I can define only I2S_1, which is the only one needed for x_nucleo_iks02a1

erwango replied:
> If I put it in board I should enable all 5 I2S, which is not good.

I don't see why. You can just enable I2S_1. This is what we are doing for other peripherals.

>  now it is good that the shield part can overload the defconfig as well

I get this point. Yet, this is in line with what you are doing when enabling setting i2s1 as arduino_i2s and enabling it.
This is just a default, conditional, configuration and this is the same we're doing with other peripherals like i2c and spi. I don't see a reason why it should be different for i2s.

avisconti replied:
OK. What happens in case a different shield requires I2S_2 or I2S_5 for example?

erwango replied:
You'll not use arduino_i2s in this case.
Shields are essentially defined by the connector they use. You're using an arduino shield, connected to what is defined as arduino_i2s, which is I2S_1 in the nucleo_f411re case (can't be changed, this is defined by a fixed set of pins).
Another shield might use another connector, or might use different pins to access SPI, but then you'll have to define other connector on board side.

avisconti replied:
OK, got your point! Actually I misunderstood completely your point until now.

avisconti commented on boards/shields/x_nucleo_iks02a1/boards/nucleo_f411re.defconfig:
```diff
@@ -0,0 +1,30 @@
+	default 192
+
+config I2S_STM32_PLLI2S_PLLR
+	default 3
```
Well again I don't think so. The nucleo_f441re does not have any audio capability. THis clock configuration strictly depends from the pdm microsoft, which requires a 128MHz I2S clock (2MHz x 64, due to oversampling factor)

erwango replied:
Ok, but this depends on the source clock (HSE/HSI), which is configured in board. So if this setting is changed on board, these settings are not valid anymore.
So at least we should add a comment to document this assumption.

avisconti replied:
sure, I can add an exhaustive comment here

avisconti committed:
soc: arm: stm32: Enable I2S_STM32 in the common part

Enable I2S_STM32 in the soc common part, so it will no
longer be required in board default configuration.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers/sensors: ism330dhcx: use endop bit in shub wait_completed

Use the endop bit in the status register (instead of using a k_sleep)
to wait until the current shub operation is completed.

Please note that the recent changes to k_sleep() API was also
breaking the compilation.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
Pushed a new version of this PR with following stuff:

1. Fixed an issue in ism330dhcx driver which was not building due to k_sleep() API recent changes.
    Moreover, it is a better way because it polls a h/w status bit instead of using just a sleep

2. Moved  I2S_STM32 in soc

3. Moved some configurations from shield to board

4. Added comments in files and improved board/sample documentation.

avisconti commented:
I came back with the idea of having a separate overlay file for mic. THis because otherwise I would not be able to compile standard for f401re (for example) and all the other boards athat are not arduino_i2s.

Re-tested and re-pushed

erwango commented on samples/shields/x_nucleo_iks02a1/sensorhub/src/main.c:
```diff
@@ -0,0 +1,252 @@
+	struct sensor_value accel1[3], accel2[3];
+	struct sensor_value gyro[3];
+	struct sensor_value magn[3];
+	struct device *iis2dlpc = device_get_binding(DT_INST_0_ST_IIS2DLPC_LABEL);
```
DT_INST_0_ST_IIS2DLPC_LABEL > DT_LABEL(DT_INST(0, st_iis2dlpc))
Apply everywehre

avisconti replied:
ok, done both

avisconti commented on boards/arm/nucleo_f411re/pinmux.c:
```diff
@@ -34,14 +34,14 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB4, STM32F4_PINMUX_FUNC_PB4_I2C3_SDA},
 #endif /* CONFIG_I2C_3 */
-#ifdef CONFIG_SPI_1
+#if defined(CONFIG_SPI_1) || defined(CONFIG_I2S_1)
```
Sure, I've done it.
But there are a couple of images, for Arduino and Morpho headers, that must be changed as well, reflecting that SPI1 and I2S1 are connected together. I do not have the source file though.

erwango replied:
Nvm for the images. These are really too much work to maintain and other boards don't have them. Maybe we should delete them.

avisconti replied:
Just tell me. If you want I can delete them and resubmit a new PR

avisconti commented:
Any other feedback guys?

avisconti commented:
@galak 
Any further comment on this PR? As you saw I'm already using it to cross verify other commits as well, like the STM32 DMA patch series from @erwango team.

avisconti commented:
Resolved conflicts and rebased. :(
Hope this PR could be merged soon.

avisconti commented:
@erwango @galak 
JUst to be 100% sure, can you possibly check what I just did in boards/arm/nucleo_f411re/pinmux.c
as trying to solve the conflicts? I think it is right, but just in case...

galak reviewed (changes requested):
Can you change all:

`include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)`

to:

`find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`

as per commit 407b49b35c111042ce474785bb88760322b26f32

avisconti commented:
> Can you change all:
> 
> `include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)`
> 
> to:
> 
> `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`
> 
> as per commit [407b49b](https://github.com/zephyrproject-rtos/zephyr/commit/407b49b35c111042ce474785bb88760322b26f32)

ack

avisconti commented:
@galak 
Can you check if I did everything you requested? Thx!

avisconti commented:
@galak @erwango @MaureenHelm 
Is it possible to have this PR merged into the upcoming release? 
I think it I addressed all the stuff here.
Thx!

erwango commented on boards/arm/nucleo_f411re/Kconfig.defconfig:
```diff
@@ -16,4 +16,11 @@ config SPI_STM32_INTERRUPT
 	default y
 	depends on SPI
 
+if I2S
+
+config I2S_1
+	default y
+
+endif # I2S
```
Please remove this.

avisconti replied:
> Please remove this.

@erwango 
Removed, re-tested, re-pushed.
Thanks again!

erwango reviewed (changes requested):
One minor comment following recent changes on master

MaureenHelm commented on drivers/sensor/ism330dhcx/ism330dhcx_shub.c:
```diff
@@ -419,10 +419,11 @@ static struct ism330dhcx_shub_slist {
-	freq = (data->accel_freq == 0) ? 26 : data->accel_freq;
-	k_sleep((2000U / freq) + 1);
+	do {
+		ism330dhcx_sh_status_get(data->ctx, &status);
```
are you sure you want to busy-wait instead of sleeping here? how long does this operation typically take?

avisconti replied:
maybe a couple of msec.
I can add a k_msleep(1), so that it put the thread to sleep after each iteration.
Would you agree?

avisconti replied:
But I would keep the status reading check to know when op is done.

MaureenHelm replied:
Yes

avisconti replied:
done, thx

MaureenHelm commented on boards/shields/x_nucleo_iks02a1/doc/index.rst:
```diff
@@ -0,0 +1,105 @@
+fitting the requirements of industrial applications.
+It is equipped with Arduino UNO R3 connector layout, and
+is designed around the ISM330DHCX 3-axis accelerometer and 3-axis gyroscope,
+the IIS2MDC 3-axis magnetometer, the IIS2DLPC 3-axis accelerometer adn
```
typo: s/adn/and/

avisconti replied:
ack

avisconti commented:
fixed, tested, re-pushed

avisconti commented:
Fixed merging issues.
Re-based to master and re-pushed.
